File: adapters/macos/Sources/kakao-macos-bridge/DumpTree.swift
import Foundation
import ApplicationServices

/// `kakao-macos-bridge --dump-tree` — serialize the live KakaoTalk
/// accessibility tree to `FixtureNode` JSON on stdout. Used to capture
/// per-version fixtures for the parser tests. Not part of the contract.
///
/// KakaoTalk Mac uses a separate window for the chat list and for each open
/// conversation, so this dumps EVERY window (via kAXWindowsAttribute, which is
/// more complete than kAXChildrenAttribute on the app element) and reports the
/// window titles + which one is main/focused on stderr.
enum DumpTree {
    static func run() {
        guard let running = KakaoApp.running() else {
            FileHandle.standardError.write(Data("KakaoTalk is not running\n".utf8))
            exit(1)
        }
        guard AX.isTrusted() else {
            FileHandle.standardError.write(Data("accessibility permission not granted\n".utf8))
            exit(1)
        }

        let appRaw = AX.app(pid: running.pid)
        AX.setMessagingTimeout(appRaw, seconds: 2.0)

        // kAXWindowsAttribute is the canonical list but goes empty for KakaoTalk
        // when windows are minimized / the app is not frontmost. Fall back to
        // the app element's direct children filtered to windows.
        var windows = AX.elements(appRaw, kAXWindowsAttribute as String)
        var windowSource = "kAXWindowsAttribute"
        if windows.isEmpty {
            windows = AX.elements(appRaw, kAXChildrenAttribute as String)
                .filter { AX.string($0, kAXRoleAttribute as String) == "AXWindow" }
            windowSource = "kAXChildrenAttribute (fallback)"
        }

        var summary = "KakaoTalk \(KakaoApp.version(of: running) ?? "?") — "
            + "\(windows.count) window(s) via \(windowSource):\n"
        var dumped: [FixtureNode] = []
        for (i, w) in windows.enumerated() {
            let el = AXElement(raw: w)
            let title = el.title ?? "(no title)"
            let sub = el.subrole ?? "-"
            let childRoles = Set(el.children.map { $0.role }).sorted().joined(separator: ",")
            summary += "  [\(i)] \(sub)  title=\(title)  childRoles={\(childRoles)}\n"
            dumped.append(snapshot(el, depth: 0, maxDepth: 60))
        }
        FileHandle.standardError.write(Data(summary.utf8))

        // Also include whatever the app element exposes directly, for comparison.
        let root = FixtureNode(
            role: "AXApplication",
            title: "KakaoTalk",
            children: dumped
        )
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .withoutEscapingSlashes, .sortedKeys]
        if let data = try? encoder.encode(root) {
            FileHandle.standardOutput.write(data)
            FileHandle.standardOutput.write(Data("\n".utf8))
        }
    }

    static func snapshot(_ node: UINode, depth: Int, maxDepth: Int) -> FixtureNode {
        let kids: [FixtureNode] = depth >= maxDepth
            ? []
            : node.children.map { snapshot($0, depth: depth + 1, maxDepth: maxDepth) }
        return FixtureNode(
            role: node.role,
            subrole: node.subrole,
            title: node.title,
            value: node.value,
            descriptionText: node.descriptionText,
            identifier: node.identifier,
            children: kids
        )
    }
}

File: adapters/windows/src/dump_tree.rs
//! `kakao-windows-bridge --dump-tree` — serialize KakaoTalk's UIA tree to
//! `FixtureNode` JSON on stdout, with a per-window summary on stderr. Used to
//! capture per-version fixtures. Not part of the contract.

#![cfg(windows)]

use crate::kakao_app;
use crate::node::FixtureNode;
use crate::uia::{self, Uia};

pub fn run() -> ! {
    let Some(running) = kakao_app::running() else {
        eprintln!("KakaoTalk is not running");
        std::process::exit(1);
    };
    let uia = match Uia::new() {
        Ok(u) => u,
        Err(e) => {
            eprintln!("UIA init failed: {:?}", e.code);
            std::process::exit(1);
        }
    };
    let ws = uia::windows_of(&uia, running.pid).unwrap_or_default();

    let mut summary = format!(
        "KakaoTalk {} — {} window(s):\n",
        kakao_app::version(&running.exe_path).unwrap_or_else(|| "?".into()),
        ws.len()
    );
    let mut dumped = Vec::new();
    for (i, w) in ws.iter().enumerate() {
        summary.push_str(&format!(
            "  [{i}] class={:?} name={:?}\n",
            w.class_name().unwrap_or_default(),
            w.name().unwrap_or_default()
        ));
        dumped.push(w.snapshot(&uia, 60));
    }
    eprint!("{summary}");

    let root = FixtureNode {
        control_type: "Pane".into(),
        name: Some("root".into()),
        automation_id: None,
        value: None,
        class_name: None,
        bounding_left: None,
        children: dumped,
    };
    match serde_json::to_string_pretty(&root) {
        Ok(s) => println!("{s}"),
        Err(e) => eprintln!("serialize failed: {e}"),
    }
    std::process::exit(0)
}
